src/Protocols/Wishbone/Standard/Hedgehog.hs

```diff
@@ -25,63 +27,184 @@ import Protocols hiding (stallC, circuit)
+--    - response data respects the 'busSelect' signal
+--  - A write request is answered with only an ACK/ERR/RETRY and no defined data
+--    - components might want to reply with data
+--
+
+type CommonSenseValidationError = String
+
+data CommonSenseValidationState
+  = CSVSQuiet
+  | CSVSInCycleNoStrobe
+  | CSVSReadCycle
+  | CSVSWriteCycle
+  deriving (Eq, Show)
+
+nextStateCommonSense ::
+  forall a addressWidth.
+  (BitPack a) =>
+  CommonSenseValidationState ->
+  WishboneM2S addressWidth (BitSize a `DivRU` 8) a ->
+  WishboneS2M a ->
+  Either CommonSenseValidationError (Bool, CommonSenseValidationState)
+--                                    ^ go to next cycle
+nextStateCommonSense _ _ s2m
+  | P.length (filter ($ s2m) [acknowledge, err, retry]) > 1 =
+    Left "More than one termination signal asserted"
+nextStateCommonSense CSVSQuiet WishboneM2S {..} s2m
+  | busCycle && not strobe = Right (True, CSVSInCycleNoStrobe)
+  | busCycle && strobe && writeEnable = Right (False, CSVSWriteCycle)
+  | busCycle && strobe && not writeEnable = Right (False, CSVSReadCycle)
+  | hasTerminateFlag s2m = Left "Termination signals outside of a bus cycle"
+  | otherwise = Right (True, CSVSQuiet)
+nextStateCommonSense CSVSInCycleNoStrobe WishboneM2S {..} _
+  | not busCycle = Right (True, CSVSQuiet)
+  | not strobe = Right (True, CSVSInCycleNoStrobe)
+  | writeEnable = Right (False, CSVSWriteCycle)
+  | not writeEnable = Right (False, CSVSReadCycle)
+  | otherwise = C.error "Should not happen"
+nextStateCommonSense CSVSReadCycle m2s@WishboneM2S {..} s2m@WishboneS2M {..}
+  | not busCycle = Right (True, CSVSQuiet)
+  | not strobe = Right (True, CSVSInCycleNoStrobe)
+  | acknowledge =
+    if responseValid m2s s2m
+      then Right (True, CSVSQuiet)
+      else Left "read-response does not respect SEL"
+  | err || retry = Right (True, CSVSQuiet)
+  | writeEnable = Left "asserted WE while in a read cycle"
+  | otherwise = Right (True, CSVSReadCycle)
+  where
+    responseValid WishboneM2S {busSelect=sel} WishboneS2M {readData=dat} = selectValidData sel dat
+nextStateCommonSense CSVSWriteCycle m2s@WishboneM2S {..} s2m@WishboneS2M {..}
+  | not busCycle = Right (True, CSVSQuiet)
+  | not strobe = Right (True, CSVSInCycleNoStrobe)
+  | not writeEnable = Left "deasserted WE while in a write cycle"
+  | not $ requestValid m2s = Left "write request does not respect SEL"
+  | err || retry = Right (True, CSVSQuiet)
+  | acknowledge =
+    if responseValid s2m
+      then Right (True, CSVSQuiet)
+      else Left "write-response contains defined read-data"
+  | otherwise = Right (True, CSVSWriteCycle)
+  where
+    requestValid WishboneM2S {busSelect=sel, writeData=dat} = selectValidData sel dat
+    responseValid WishboneS2M {readData=dat0} =
+      let dat1 = bitCoerce dat0
+          numBits = natVal @_ @BitVector dat1
+          mask = 2 P.^ numBits - 1
+       in unsafeMask dat1 == mask
+
+selectValidData ::
+  forall a.
```

Comment:
Can you add documentation here too? Even though the user will never see it, developers will.

Reply:
Also, if you function is going to produce bottoms, add `HasCallStack`.